### ISSUE TITLE
Flux.gradient returns dict of params and nothing

### ISSUE THREAD

**Issue by sanderbboisen** [FluxML/Flux.jl#1713]
Hi everyone,
I am fairly new to Julia and Flux and I am trying to implement a WGAN-variant that I have implemented in Keras before. Full code for the issue can be found [here](https://gist.github.com/sanderbboisen/8e667b78348587169a4ff90f2c5dbe7c).

I keep getting gradients that consist of pairs of ("SOME_KEY", "Nothing") which in the end leads to Flux not updating the model when `Flux.optimise.update!` is called.
These are the problematic lines: 
```
... 
grads = Flux.gradient(ps) do
        loss = wasserstein_g_loss(gen_data)
end;
Flux.update!(opt_gen, ps, grads)
...
```
The alternative
```
...
 loss, grads = Flux.pullback(ps) do # Do all relevant calculations within pullback
        gen_data = gen(gen_in)
        gen_norm = gradient_normalization(crit, gen_data)
        loss = wasserstein_g_loss(gen_norm)
    end;
    grads = grads(1f0)
    Flux.Optimize.update!(opt_gen, ps, grads)
...
```
Does not work either.

I have tried a bunch of variations of calling `Flux.gradient` and `Flux.pullback` but nothing seems to work for me.  Both methods work for me in other contexts, and I am reaching the end of my rope here. 

If you read through this post and took a look at the code, I thank you kindly for your time.

Cheers!

**Comment by ToucheSir** [comment]
I'm going to close this in favour of commenting on the older [Discourse topic](https://discourse.julialang.org/t/flux-gradient-returns-dict-of-param-and-nothing/68457/2) it was copied from, feel free to continue there!

**Comment by DhairyaLGandhi** [comment]
Your second approach is close! You would want to have the code you want gradients for inside the `gradient` context.
